### PR TITLE
[build] Add vcpkg submodule and use protobuf

### PR DESCRIPTION
**Pull request by townsend2010**

### CODE REVIEW

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -80,6 +80,15 @@ jobs:
         [ ! -f ${PATCH_PREFIX}.patch ] || patch -p1 --no-backup-if-mismatch < ${PATCH_PREFIX}.patch
         [ ! -f ${PATCH_PREFIX}-${{ matrix.build-type }}.patch ] || patch -p1 --no-backup-if-mismatch < ${PATCH_PREFIX}-${{ matrix.build-type }}.patch
 
+    - name: Set up vcpkg
+      uses: lukka/run-vcpkg@v7.1
+      with:
+        setupOnly: true
+        vcpkgDirectory: '${{ github.workspace }}/3rd-party/vcpkg'
+        appendedCacheKey: ${{ hashFiles( '**/vcpkg_manifest/vcpkg.json' ) }}
+        vcpkgTriplet: ${{ matrix.triplet }}
```
Comment (Saviq) [inline .github/workflows/linux.yml:89]: This ends up empty, as is unset in the matrix above.

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -80,6 +80,15 @@ jobs:
         [ ! -f ${PATCH_PREFIX}.patch ] || patch -p1 --no-backup-if-mismatch < ${PATCH_PREFIX}.patch
         [ ! -f ${PATCH_PREFIX}-${{ matrix.build-type }}.patch ] || patch -p1 --no-backup-if-mismatch < ${PATCH_PREFIX}-${{ matrix.build-type }}.patch
 
+    - name: Set up vcpkg
+      uses: lukka/run-vcpkg@v7.1
+      with:
+        setupOnly: true
+        vcpkgDirectory: '${{ github.workspace }}/3rd-party/vcpkg'
+        appendedCacheKey: ${{ hashFiles( '**/vcpkg_manifest/vcpkg.json' ) }}
```
Comment (Saviq) [inline .github/workflows/linux.yml:88]: This should probably be `vcpkg.json`, as we have no `vcpkg_manifest`?

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -80,6 +81,20 @@ jobs:
         [ ! -f ${PATCH_PREFIX}.patch ] || patch -p1 --no-backup-if-mismatch < ${PATCH_PREFIX}.patch
         [ ! -f ${PATCH_PREFIX}-${{ matrix.build-type }}.patch ] || patch -p1 --no-backup-if-mismatch < ${PATCH_PREFIX}-${{ matrix.build-type }}.patch
 
+    - name: Set up vcpkg archive cache
+      id: setup-vcpkg-cache
```
Comment (Saviq) [inline .github/workflows/linux.yml:85]:
```suggestion
```

File: .github/workflows/linux.yml [hunk 4]
```diff
@@ -80,6 +81,20 @@ jobs:
...
+      run: |
+        mkdir --parents ${{ env.VCPKG_INSTALL_CACHE }}
+        sg lxd -c '/snap/bin/lxc profile device add default vcpkg disk source=${{ env.VCPKG_INSTALL_CACHE }} path=/root/parts/multipass/build/vcpkg_installed'
```
Comment (Saviq) [inline .github/workflows/linux.yml:88]:
```suggestion
        sg lxd -c '/snap/bin/lxc profile device add default vcpkg disk source=${VCPKG_INSTALL_CACHE} path=/root/parts/multipass/build/vcpkg_installed'
```

File: .github/workflows/linux.yml [hunk 5]
```diff
@@ -80,6 +81,20 @@ jobs:
...
+
+    - name: Set up vcpkg
+      uses: lukka/run-vcpkg@v7.1
+      with:
+        setupOnly: true
+        vcpkgDirectory: '${{ github.workspace }}/3rd-party/vcpkg'
+        appendedCacheKey: ${{ runner.os }}-${{ matrix.build-type }}-${{ hashFiles( '**/vcpkg.json' ) }}
+        additionalCachedPaths: ${{ env.VCPKG_INSTALL_CACHE }}
```
Comment (Saviq) [inline .github/workflows/linux.yml:96]: Does this make sense? Do we need vcpkg outside of the LXD container?

If not, you can use the cache action directly:
```yaml
    - name: CCache
      uses: actions/cache@v2
      with:
        key: vcpkg-${{ runner.os }}-${{ matrix.build-type }}-${{ hashFiles( '**/vcpkg.json' ) }}
        path: ${{ env.VCPKG_INSTALL_CACHE }}
```

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -94,6 +94,10 @@ parts:
     - pkg-config
     - qtbase5-dev
     - qtbase5-dev-tools
+    - curl
+    - zip
+    - unzip
+    - tar
```
Comment (Saviq) [inline snap/snapcraft.yaml:100]: vcpkg needs all these? Please sort.

**Review by Saviq — CHANGES_REQUESTED** [review]

File: .github/workflows/linux.yml (hunk 4 above)
Comment (townsend2010) [reply .github/workflows/linux.yml:88]: I'm not I understand why the environment variable itself can be used here, but the previous `mkdir` command needs the context expansion?

File: snap/snapcraft.yaml (hunk 6 above)
Comment (townsend2010) [reply snap/snapcraft.yaml:100]: Yes and will do.

File: .github/workflows/linux.yml (hunk 5 above)
Comment (townsend2010) [reply .github/workflows/linux.yml:95]: As discussed in chat, the `run-vcpkg` action uses the `cache` action internally.

File: .github/workflows/linux.yml (hunk 4 above)
Comment (Saviq) [reply .github/workflows/linux.yml:87]: It doesn't need it, I could've updated it, too.

`${{ }}` are processed before it ever gets to BASH.

File: .github/workflows/linux.yml (hunk 5 above)
Comment (Saviq) [reply .github/workflows/linux.yml:95]: More importantly, the `vcpkg` executable gets built in this step, and is shipped along with the source into the snap. Which I dislike a lot ;)

If there's a way to change where vcpkg gets built, in a way that makes it not pollute the source, but rather it getting shipped along with `vcpkg_installed`, that would be nice.

File: .github/workflows/linux.yml (hunk 4 above)
Comment (townsend2010) [reply .github/workflows/linux.yml:87]: Ok, good to know.  I'll fix these up.

File: .github/workflows/linux.yml (hunk 5 above)
Comment (townsend2010) [reply .github/workflows/linux.yml:95]: Ok, I see your point and will see if there is a way to massage `vcpkg` to not build directly in its submodule directory.  If so, then I may be able to dump the `run-vcpkg` action and just use the `cache action` along with any of our own logic.

File: .github/workflows/linux.yml (hunk 4 above)
Comment (townsend2010) [reply .github/workflows/linux.yml]: Done

**Comment by townsend2010** [comment]
Ok, got this in a state where I can come back to it later, but going to shelve it for now.

**Comment by townsend2010** [comment]
Killing this as it's not worth the effort.

**Pull request closed without merging**